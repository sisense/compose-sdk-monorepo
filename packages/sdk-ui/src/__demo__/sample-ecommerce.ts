import { measureFactory } from '@sisense/sdk-data';

import { Commerce, Brand, Country, Category, DataSource } from './sample-ecommerce-autogenerated';

export { Commerce, Brand, Category, Country, DataSource };

/** adding predefined set of measures */
export const Measures = {
  SumCost: measureFactory.sum(Commerce.Cost).format('0,0$'),
  AvgCost: measureFactory.average(Commerce.Cost).format('0,0$'),
  SumRevenue: measureFactory.sum(Commerce.Revenue, 'Total Revenue').format('0,0$'),
  Quantity: measureFactory.sum(Commerce.Quantity, 'Total Quantity').format('0,0'),
  CountBrands: measureFactory.countDistinct(Commerce.BrandID).format('0,0'),
};
