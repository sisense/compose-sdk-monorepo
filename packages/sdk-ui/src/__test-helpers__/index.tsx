/**
 * Common test utilities
 */
import { PropsWithChildren, ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';
import type { Cell, Data } from '@ethings-os/sdk-data';
import isObject from 'lodash-es/isObject';
import userEvent from '@testing-library/user-event';
import { SisenseContext, SisenseContextPayload } from '../sisense-context/sisense-context';
import { Authenticator, HttpClient } from '@ethings-os/sdk-rest-client';
import { ClientApplication } from '../app/client-application';
import { DeepPartial } from 'ts-essentials';
import { useTranslation } from 'react-i18next';
import { translation } from '../translation/resources/en';
import { CustomWidgetsProvider } from '../custom-widgets-provider';
import { MenuProvider } from '@/common/components/menu/menu-provider';

type UserSetupFn = (typeof userEvent)['setup'];
interface SetupResult extends RenderResult {
  user: ReturnType<UserSetupFn>;
}
/**
 * This helper method combines the userEvent.setup() and render() calls at the
 * start of a unit test using @testing-library/react.
 *
 * Based on recommendation by: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
 *
 * @param jsx - React component instance to test, passed directly to render()
 * @param useFakeTimers - true if Vitest fake timers are enabled
 * @returns
 */
export function setup(jsx: ReactElement, useFakeTimers = false): SetupResult {
  return {
    // Need to set "advanceTimers" if we're using Vitest's fake timers.
    // The solution in https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841
    // is for Jest, but this seems to work for Vitest as well.
    user: userEvent.setup(
      useFakeTimers ? { advanceTimers: (ms) => vi.advanceTimersByTime(ms) } : undefined,
    ),
    ...render(jsx),
  };
}

export const MockedHighchartsReact = ({
  options,
}: {
  /**
   * Highcharts chart configuration object.
   * Please refer to the Highcharts (API documentation)[https://api.highcharts.com/highcharts/].
   */
  options: object;
}) => {
  return <div>{JSON.stringify(options)}</div>;
};

type MockedSisenseContextProviderProps = PropsWithChildren<
  DeepPartial<Omit<SisenseContextPayload, 'app'>>
>;

const mockHttpClient = new HttpClient(
  'http://mock-url/sometenant?someparam=true',
  {} as Authenticator,
  'mock-env',
);
export const MockedSisenseContextProvider = ({
  children,
  tracking = {
    enabled: true,
  },
  isInitialized = true,
  errorBoundary,
}: MockedSisenseContextProviderProps) => {
  return (
    <SisenseContext.Provider
      value={{
        app: {
          httpClient: mockHttpClient,
        } as ClientApplication,
        tracking: {
          enabled: tracking.enabled ?? true,
          packageName: tracking.packageName ?? 'sdk-ui',
        },
        isInitialized,
        errorBoundary: { showErrorBox: true, ...errorBoundary },
      }}
    >
      <MenuProvider>
        <CustomWidgetsProvider>{children}</CustomWidgetsProvider>
      </MenuProvider>
    </SisenseContext.Provider>
  );
};

/**
 * Extends the provided data by blurring specified rows.
 *
 * @param {Data} data - The original data object.
 * @param {number[]} blurRowsIndexes - An array of row indexes to be blurred.
 * @returns {Data} The data object with blurred rows.
 */
export const withBlurredRows = (data: Data, blurRowsIndexes: number[]) => ({
  ...data,
  rows: data.rows.map((row, rowIndex) =>
    row.map((value) => {
      const cell: Cell = isObject(value) ? value : { data: value };

      if (blurRowsIndexes.includes(rowIndex)) {
        return { ...cell, blur: true };
      }
      return cell;
    }),
  ),
});

/**
 * Mocking the i18next translation functionality for tests.
 */
export const setupI18nMock = () => {
  vi.mock('react-i18next', async () => {
    const original: { useTranslation: typeof useTranslation } = await vi.importActual(
      'react-i18next',
    );

    // Helper function to get nested translations safely
    const getNestedTranslation = (key: string): string | undefined => {
      return key.split('.').reduce((accumulator: any, currentKey: string) => {
        if (accumulator && typeof accumulator === 'object' && currentKey in accumulator) {
          return accumulator[currentKey];
        }
        return undefined;
      }, translation) as string | undefined;
    };

    return {
      ...original,
      useTranslation: () => ({
        ...original.useTranslation(),
        t: vi.fn((key, options?: Record<string, any>) => {
          const translatedValue = getNestedTranslation(key);
          if (translatedValue && options) {
            // Replace variables in the translated string
            return translatedValue.replace(/{{(.*?)}}/g, (_, varName) => {
              const trimmedVarName = varName.trim();
              return options[trimmedVarName] || `{{${trimmedVarName}}}`; // Fallback if variable is not found
            });
          }
          return translatedValue ?? key; // Fallback to the key if translation is not found
        }),
      }),
    };
  });
};
