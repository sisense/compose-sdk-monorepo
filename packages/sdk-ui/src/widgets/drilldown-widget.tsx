/* eslint-disable max-lines-per-function */

import { useState, useCallback, useMemo } from 'react';
import { Attribute } from '@sisense/sdk-data';
import { DataPoint, MenuPosition, MenuItemSection } from '../types';
import { ContextMenu } from './common/context-menu';
import { DrilldownBreadcrumbs } from './common/drilldown-breadcrumbs';
import { useCustomDrilldown } from './common/custom-drilldown';
import { DrilldownWidgetProps } from '../props';

/**
 * React component designed to add drilldown functionality to any type of chart
 *
 * It acts as a wrapper around a given chart component, enhancing it with drilldown capabilities
 *
 * The widget offers several features including:
 * - A context menu for initiating drilldown actions (can be provided as a custom component)
 * - Breadcrumbs that not only allow for drilldown selection slicing but also
 * provide an option to clear the selection (can be provided as a custom component)
 * - Filters specifically created for drilldown operation
 * - An option to navigate to the next drilldown dimension
 *
 * When an `initialDimension` is specified, the `drilldownDimension` will automatically inherit its value,
 * even before any points on the chart are selected.
 * This allows for complete control over the chart's dimensions to be handed over to the DrilldownWidget
 *
 * @example
 * Example of using the `DrilldownWidget` component to
 * plot a custom React component that uses the `ExecuteQuery` component to
 * query the `Sample ECommerce` data source hosted in a Sisense instance.
 * ```tsx
 * <DrilldownWidget
 *   drilldownDimensions={[DM.Commerce.AgeRange, DM.Commerce.Gender, DM.Commerce.Condition]}
 *   initialDimension={DM.Category.Category}
 * >
 *   {({ drilldownFilters, drilldownDimension, onDataPointsSelected, onContextMenu }) => (
 *     <ExecuteQuery
 *       dataSource={DM.DataSource}
 *       dimensions={[drilldownDimension]}
 *       measures={measure.sum(DM.Commerce.Revenue)}
 *       filters={drilldownFilters}
 *     >
 *       {(data) => (
 *         <MyCustomChart
 *           rawData={data}
 *           onContextMenu={onContextMenu}
 *           onDataPointsSelected={onDataPointsSelected}
 *         />
 *       )}
 *     </ExecuteQuery>
 *   )}
 * </DrilldownWidget>
 * ```
 * @param props - DrilldownWidget properties
 * @returns DrilldownWidget wrapper component
 */
export const DrilldownWidget = ({
  drilldownDimensions,
  initialDimension,
  config,
  children,
}: DrilldownWidgetProps) => {
  const [selectedDataPoints, setSelectedDataPoints] = useState<DataPoint[]>([]);
  const [contextMenuPos, setContextMenuPos] = useState<null | MenuPosition>(null);

  const ContextMenuComponent = config?.contextMenuComponent ?? ContextMenu;
  const BreadcrumbsComponent = config?.breadcrumbsComponent ?? DrilldownBreadcrumbs;

  const {
    selectDrilldown,
    sliceDrilldownSelections,
    clearDrilldownSelections,
    availableDrilldowns,
    drilldownFilters,
    drilldownFiltersDisplayValues,
    drilldownDimension,
  } = useCustomDrilldown({
    drilldownDimensions,
    initialDimension,
  });

  const breadcrumbs = useMemo(() => {
    return (
      drilldownDimension && (
        <BreadcrumbsComponent
          filtersDisplayValues={drilldownFiltersDisplayValues}
          currentDimension={drilldownDimension}
          clearDrilldownSelections={clearDrilldownSelections}
          sliceDrilldownSelections={sliceDrilldownSelections}
        />
      )
    );
  }, [
    BreadcrumbsComponent,
    clearDrilldownSelections,
    drilldownDimension,
    drilldownFiltersDisplayValues,
    sliceDrilldownSelections,
  ]);

  const openContextMenu = (menuPos: { top: number; left: number }) => {
    setContextMenuPos(menuPos);
  };

  const closeContextMenu = useCallback(() => {
    setSelectedDataPoints([]);
    setContextMenuPos(null);
  }, [setSelectedDataPoints]);

  const onMenuDrilldownClick = useCallback(
    (nextDimension: Attribute) => {
      selectDrilldown(selectedDataPoints, nextDimension);
    },
    [selectDrilldown, selectedDataPoints],
  );

  const onDataPointsSelected = useCallback(
    (points: DataPoint[]) => {
      setSelectedDataPoints(points);
    },
    [setSelectedDataPoints],
  );

  const drilldownMenuItems: MenuItemSection[] = useMemo(
    () => [
      ...(drilldownDimension ? [{ sectionTitle: drilldownDimension?.name }] : []),
      {
        sectionTitle: 'Drill',
        items: availableDrilldowns.map((nextDimension) => ({
          caption: nextDimension.name,
          onClick: () => onMenuDrilldownClick(nextDimension),
        })),
      },
    ],
    [drilldownDimension, availableDrilldowns, onMenuDrilldownClick],
  );

  return (
    <>
      <ContextMenuComponent
        position={contextMenuPos}
        itemSections={drilldownMenuItems}
        closeContextMenu={closeContextMenu}
      />
      {drilldownDimension && !config?.isBreadcrumbsDetached && breadcrumbs}
      {children({
        drilldownFilters,
        drilldownDimension,
        onDataPointsSelected,
        onContextMenu: openContextMenu,
        breadcrumbsComponent: config?.isBreadcrumbsDetached ? breadcrumbs : undefined,
      })}
    </>
  );
};
