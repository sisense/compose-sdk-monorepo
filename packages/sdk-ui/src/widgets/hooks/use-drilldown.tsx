/* eslint-disable @typescript-eslint/no-use-before-define */
import { useCallback, useMemo } from 'react';
import { useTranslation } from 'react-i18next';

import { type TFunction } from '@sisense/sdk-common';
import { Attribute, Column, MetadataTypes } from '@sisense/sdk-data';
import partition from 'lodash-es/partition';
import uniq from 'lodash-es/uniq';

import { SELECTION_TITLE_MAXIMUM_ITEMS } from '@/common-filters/selection-utils.js';
import { MenuIds, MenuSectionIds } from '@/common/components/menu/menu-ids.js';
import { OpenMenuFn } from '@/common/components/menu/types.js';
import { Hierarchy } from '@/models';
import { isSameAttribute } from '@/utils/filters.js';

import {
  DataPoint,
  DrilldownSelection,
  MenuItemSection,
  MenuPosition,
  StyledColumn,
} from '../../types.js';
import { DrilldownBreadcrumbs } from '../common/drilldown-breadcrumbs/drilldown-breadcrumbs.js';
import '../common/drilldown.scss';
import {
  getDisplayMemberNameFromDataPoint,
  useDrilldownCore,
} from '../common/use-drilldown-core.js';

type UseDrilldownParams = {
  initialDimension: Column | StyledColumn;
  drilldownPaths?: (Attribute | Hierarchy)[];
  drilldownSelections?: DrilldownSelection[];
  /**
   * todo: make it optional when we will have a public `MenuProvider`.
   * Without provided function, it should open internal menu with the help of `useMenu` hook
   */
  openMenu: OpenMenuFn;
  onDrilldownSelectionsChange?: (selections: DrilldownSelection[]) => void;
};

export const useDrilldown = ({
  initialDimension,
  drilldownPaths,
  drilldownSelections,
  openMenu,
  onDrilldownSelectionsChange,
}: UseDrilldownParams) => {
  const { t: translate } = useTranslation();

  const {
    availableDrilldownPaths,
    drilldownFilters,
    drilldownFiltersDisplayValues,
    drilldownDimension,
    selectDrilldown,
    sliceDrilldownSelections,
    clearDrilldownSelections,
  } = useDrilldownCore({
    initialDimension,
    drilldownPaths,
    drilldownSelections,
    onDrilldownSelectionsChange,
  });

  const openDrilldownMenu = useCallback(
    (position: MenuPosition, points: DataPoint[]) => {
      const menuItems = [
        getSelectionTitleMenuItem(points, drilldownDimension),
        getDrilldownMenuItems(
          availableDrilldownPaths,
          drilldownDimension,
          (nextDimension: Attribute, hierarchy?: Hierarchy) => {
            selectDrilldown(points, nextDimension, hierarchy);
          },
          translate,
        ),
      ];

      openMenu({ id: MenuIds.WIDGET_POINTS_DRILLDOWN, position, itemSections: menuItems });
    },
    [drilldownDimension, availableDrilldownPaths, translate, selectDrilldown, openMenu],
  );

  const breadcrumbs = useMemo(() => {
    return (
      drilldownDimension && (
        <DrilldownBreadcrumbs
          filtersDisplayValues={drilldownFiltersDisplayValues}
          currentDimension={drilldownDimension}
          clearDrilldownSelections={clearDrilldownSelections}
          sliceDrilldownSelections={sliceDrilldownSelections}
        />
      )
    );
  }, [
    clearDrilldownSelections,
    drilldownDimension,
    drilldownFiltersDisplayValues,
    sliceDrilldownSelections,
  ]);

  return {
    drilldownDimension,
    drilldownFilters,
    breadcrumbs,
    openDrilldownMenu,
  };
};

/** @internal */
export function getSelectionTitleMenuItem(
  points: DataPoint[],
  drilldownDimension: Attribute,
): MenuItemSection {
  const selectionNames = uniq(
    points.map(getDisplayMemberNameFromDataPoint).filter((name) => !!name),
  );

  if (selectionNames.length > SELECTION_TITLE_MAXIMUM_ITEMS) {
    return {
      id: MenuSectionIds.DRILLDOWN_CHART_POINTS_SELECTION,
      sectionTitle: drilldownDimension.name,
    };
  }

  return {
    id: MenuSectionIds.DRILLDOWN_CHART_POINTS_SELECTION,
    sectionTitle: selectionNames.join(', '),
  };
}

function getHierarchyLevelMenuItemOffset(index: number) {
  const MENU_ITEM_DEFAUT_PADDING_LEFT = 30;
  const NESTED_MENU_ITEM_OFFSET_LEFT = 12;

  return MENU_ITEM_DEFAUT_PADDING_LEFT + NESTED_MENU_ITEM_OFFSET_LEFT * index;
}

/** @internal */
export function getDrilldownMenuItems(
  availableDrilldownPaths: (Attribute | Hierarchy)[],
  drilldownDimension: Attribute,
  selectFn: (nextDimension: Attribute, hierarchy?: Hierarchy) => void,
  translate: TFunction,
): MenuItemSection {
  const [availableDrilldownDimensions, availableDrilldownHierarchies] = partition<
    Attribute | Hierarchy,
    Attribute
  >(availableDrilldownPaths, MetadataTypes.isAttribute.bind(MetadataTypes));
  const drilldownDimensionsMenuItems = availableDrilldownDimensions.map((nextDimension) => ({
    caption: nextDimension.name,
    onClick: () => selectFn(nextDimension),
  }));
  const drilldownHierarchiesMenuItems = availableDrilldownHierarchies.map((hierarchy) => {
    return {
      caption: hierarchy.title,
      subItems: [
        {
          items: hierarchy.levels.map((level, index) => ({
            caption: level.name,
            class: index === 0 ? '' : 'csdk-drilldown-hierarchy-nested-menu-item',
            style: {
              paddingLeft: `${getHierarchyLevelMenuItemOffset(index)}px`,
            },
            disabled: isSameAttribute(level, drilldownDimension),
            onClick: () => selectFn(level, hierarchy),
          })),
        },
      ],
    };
  });

  return {
    id: MenuSectionIds.DRILLDOWN_DRILL_DIRECTIONS,
    sectionTitle: translate('drilldown.drillMenuItem'),
    items: [...drilldownHierarchiesMenuItems, ...drilldownDimensionsMenuItems],
  };
}
