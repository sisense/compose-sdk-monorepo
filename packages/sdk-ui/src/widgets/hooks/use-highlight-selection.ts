import { useCallback, useMemo, useState } from 'react';

import { HighchartsOptionsInternal } from '@/chart-options-processor/chart-options-service';
import { isCartesian, isScatter } from '@/chart-options-processor/translations/types';

import {
  CartesianChartDataOptions,
  ChartDataOptions,
  ChartDataPoint,
  ChartDataPoints,
  ChartType,
  DataPoint,
  HighchartsOptions,
  ScatterDataPoint,
} from '../..';

export function useHighlightSelection({
  chartType,
  dataOptions,
  enabled,
}: {
  chartType: ChartType;
  dataOptions: ChartDataOptions;
  enabled: boolean;
}) {
  const [selectedDataPoints, setSelectedDataPoints] = useState<ChartDataPoints>([]);

  const isSelectionAllowed = useMemo(() => {
    const isSupportedChart =
      (isCartesian(chartType) &&
        (dataOptions as CartesianChartDataOptions).category?.length === 1) ||
      isScatter(chartType);
    return enabled && isSupportedChart;
  }, [enabled, chartType, dataOptions]);

  const applyPointSelections = useMemo(() => {
    if (selectedDataPoints.length === 0) {
      return (options: HighchartsOptionsInternal) =>
        ({
          ...options,
          series: options.series.map((s) => ({
            ...s,
            data: s.data.map((d) => ({ ...d, selected: false })),
          })),
        } as HighchartsOptions);
    }
    if (isScatter(chartType)) {
      return (options: HighchartsOptionsInternal) =>
        ({
          ...options,
          series: options.series.map((s) => ({
            ...s,
            data: s.data.map((d) => {
              return {
                ...d,
                selected: !(selectedDataPoints as ScatterDataPoint[]).some(
                  (point) =>
                    point.x === d.x &&
                    point.y === d.y &&
                    point.size === d.z &&
                    point.breakByPoint === d.custom?.maskedBreakByPoint &&
                    point.breakByColor === d.custom?.maskedBreakByColor,
                ),
              };
            }),
          })),
        } as HighchartsOptions);
    } else {
      const categoryValueMap = (selectedDataPoints as DataPoint[]).reduce(
        (accu, { categoryValue }) => {
          if (categoryValue) {
            accu[`${categoryValue}`] = true;
          }
          return accu;
        },
        {},
      );

      return (options: HighchartsOptionsInternal) => ({
        ...options,
        series: options.series.map((s) => ({
          ...s,
          data: s.data.map((d) =>
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            categoryValueMap[d.custom?.xValue?.[0]] ? d : { ...d, selected: true },
          ),
        })),
      });
    }
  }, [selectedDataPoints, chartType]);

  const setOnClickOutside = useMemo(
    () => (options: HighchartsOptions) =>
      ({
        ...options,
        chart: {
          ...options.chart,
          events: {
            ...options.chart?.events,
            click: () => {
              setSelectedDataPoints([]);
            },
          },
        },
      } as HighchartsOptions),
    [],
  );

  const onBeforeRender = useCallback(
    (options: HighchartsOptions) => {
      if (isSelectionAllowed) {
        options = applyPointSelections(options as HighchartsOptionsInternal) as HighchartsOptions;
        options = setOnClickOutside(options);
      }
      return options;
    },
    [applyPointSelections, setOnClickOutside, isSelectionAllowed],
  );

  const onDataPointsSelected = useCallback(
    (dataPoints: ChartDataPoints): void => {
      if (isSelectionAllowed) {
        setSelectedDataPoints(dataPoints);
      }
    },
    [isSelectionAllowed],
  );

  const onDataPointClick = useCallback(
    (dataPoint: ChartDataPoint) => {
      if (isSelectionAllowed) {
        setSelectedDataPoints([dataPoint] as ChartDataPoints);
      }
    },
    [isSelectionAllowed],
  );

  return {
    onBeforeRender,
    onDataPointsSelected,
    onDataPointClick,
  };
}
