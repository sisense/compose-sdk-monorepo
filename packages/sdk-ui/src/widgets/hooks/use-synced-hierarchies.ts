import { useMemo } from 'react';
import { Attribute, DataSource } from '@sisense/sdk-data';
import isString from 'lodash-es/isString';
import partition from 'lodash-es/partition';
import { Hierarchy, HierarchyId } from '@/models/hierarchy';
import { useGetHierarchyModelsInternal } from '@/models/hierarchy/use-get-hierarchy-models';

/**
 * Synchronizes drilldown paths with hierarchy data from a Fusion instance.
 * It loads all hierarchies provided as hierarchy IDs, and also includes any hierarchies
 * that are marked as `alwaysIncluded` in a Fusion instance.
 *
 * @internal
 */
export function useSyncedDrilldownPaths({
  attribute,
  dataSource,
  drilldownPaths = [],
  enabled,
}: {
  attribute: Attribute;
  dataSource?: DataSource;
  drilldownPaths?: (Attribute | Hierarchy | HierarchyId)[];
  enabled: boolean;
}) {
  // Separate hierarchy IDs from the plain hierarchy objects and dimensions in `drilldownPaths` list
  const [hierarchyIds, otherDrilldownPaths] = useMemo(() => {
    return partition(drilldownPaths, isString);
  }, [drilldownPaths]);

  const { hierarchies: loadedHierarchies = [] } = useGetHierarchyModelsInternal({
    enabled,
    dataSource,
    dimension: attribute,
    ids: hierarchyIds,
    alwaysIncluded: true,
  });

  return useMemo(
    () =>
      loadedHierarchies.length
        ? [...loadedHierarchies, ...otherDrilldownPaths]
        : otherDrilldownPaths,
    [loadedHierarchies, otherDrilldownPaths],
  );
}
