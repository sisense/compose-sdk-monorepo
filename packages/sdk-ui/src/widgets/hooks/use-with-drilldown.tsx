/* eslint-disable @typescript-eslint/no-use-before-define */
import { useCallback, useMemo } from 'react';
import { Filter, createAttribute } from '@sisense/sdk-data';
import { DataPoint, DrilldownSelection, ScatterDataPoint } from '../../types.js';
import { ChartWidgetProps } from '@/props.js';
import { useMenu } from '@/common/hooks/use-menu.js';
import { useDrilldown } from './use-drilldown.js';
import { useSyncedDrilldownPaths } from './use-synced-hierarchies.js';
import {
  applyDrilldownDimension,
  getDrilldownInitialDimension,
  isDrilldownApplicableToChart,
  prepareDrilldownSelectionPoints,
} from '../common/drilldown-utils.js';
import { mergeFilters } from '@/dashboard-widget/utils.js';
import { combineHandlers } from '@/utils/combine-handlers.js';

type UseWithDrilldownParams = {
  propsToExtend: ChartWidgetProps;
  onDrilldownSelectionsChange?: (selections: DrilldownSelection[]) => void;
};

export const useWithDrilldown = ({
  propsToExtend,
  onDrilldownSelectionsChange,
}: UseWithDrilldownParams) => {
  const { chartType, dataOptions, dataSource, drilldownOptions } = propsToExtend;
  const { drilldownSelections } = drilldownOptions || {};
  const { openMenu } = useMenu();

  const isDrilldownApplicable = useMemo(
    () => isDrilldownApplicableToChart(chartType, dataOptions),
    [chartType, dataOptions],
  );
  const initialDrilldownPaths = useMemo(
    () => [
      ...(drilldownOptions?.drilldownPaths || []),
      ...(drilldownOptions?.drilldownDimensions || []),
    ],
    [drilldownOptions?.drilldownPaths, drilldownOptions?.drilldownDimensions],
  );
  const initialDimension = useMemo(() => {
    const dummyAttribute = createAttribute({ name: 'dummy drilldown initial attribute' });
    return getDrilldownInitialDimension(chartType, dataOptions) || dummyAttribute;
  }, [chartType, dataOptions]);

  const drilldownPaths = useSyncedDrilldownPaths({
    attribute: initialDimension,
    dataSource: dataSource,
    drilldownPaths: initialDrilldownPaths,
    enabled: isDrilldownApplicable,
  });

  const isDrilldownEnabled = useMemo(() => {
    const hasDrilldownConfig =
      drilldownOptions?.drilldownSelections?.length ||
      drilldownOptions?.drilldownDimensions?.length ||
      drilldownPaths?.length;

    return hasDrilldownConfig && isDrilldownApplicable;
  }, [drilldownOptions, isDrilldownApplicable, drilldownPaths]);

  const { drilldownDimension, drilldownFilters, breadcrumbs, openDrilldownMenu } = useDrilldown({
    initialDimension,
    drilldownSelections,
    drilldownPaths,
    openMenu,
    onDrilldownSelectionsChange,
  });

  const drilldownOnDataPointsSelected = useCallback(
    (points: (DataPoint | ScatterDataPoint)[], event: MouseEvent) => {
      const drilldownSelectionPoints = prepareDrilldownSelectionPoints(points, event, dataOptions);
      openDrilldownMenu(
        {
          left: event.clientX,
          top: event.clientY,
        },
        drilldownSelectionPoints,
      );
    },
    [dataOptions, openDrilldownMenu],
  );

  const drilldownOnDataPointContextMenu = useCallback(
    (point: DataPoint, event: MouseEvent) => {
      const drilldownSelectionPoints = prepareDrilldownSelectionPoints([point], event, dataOptions);
      openDrilldownMenu(
        {
          left: event.clientX,
          top: event.clientY,
        },
        drilldownSelectionPoints,
      );
    },
    [dataOptions, openDrilldownMenu],
  );

  /**
   * Connecting drilldown to props:
   */
  const dataOptionsWithDrilldown = useMemo(() => {
    return applyDrilldownDimension(chartType, dataOptions, drilldownDimension);
  }, [chartType, dataOptions, drilldownDimension]);

  const filtersWithDrilldown = useMemo(() => {
    return drilldownFilters.length
      ? // todo: cover filter relations case
        mergeFilters(propsToExtend.filters as Filter[], drilldownFilters)
      : propsToExtend.filters;
  }, [propsToExtend.filters, drilldownFilters]);

  const onDataPointsSelectedWithDrilldown = useMemo(
    () => combineHandlers([drilldownOnDataPointsSelected, propsToExtend.onDataPointsSelected]),
    [drilldownOnDataPointsSelected, propsToExtend.onDataPointsSelected],
  );

  const onDataPointContextMenuWithDrilldown = useMemo(
    () => combineHandlers([drilldownOnDataPointContextMenu, propsToExtend.onDataPointContextMenu]),
    [drilldownOnDataPointContextMenu, propsToExtend.onDataPointContextMenu],
  );

  const propsWithDrilldown = {
    ...propsToExtend,
    ...(isDrilldownEnabled && {
      dataOptions: dataOptionsWithDrilldown,
      filters: filtersWithDrilldown,
      onDataPointsSelected: onDataPointsSelectedWithDrilldown,
      onDataPointContextMenu: onDataPointContextMenuWithDrilldown,
    }),
  } as ChartWidgetProps;

  return {
    propsWithDrilldown,
    isDrilldownEnabled,
    breadcrumbs,
  };
};
