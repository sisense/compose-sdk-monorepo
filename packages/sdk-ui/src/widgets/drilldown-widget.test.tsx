import { render, fireEvent } from '@testing-library/react';

import { DrilldownWidget } from './drilldown-widget';
import { useMenu } from '@/common/hooks/use-menu';
import { useSisenseContext } from '@/sisense-context/sisense-context';
import { Mock } from 'vitest';

import * as DM from '@/__test-helpers__/sample-ecommerce';

vi.mock('@/common/hooks/use-menu', () => ({
  useMenu: vi.fn(),
}));

vi.mock('@/sisense-context/sisense-context');

const mockUseMenu = useMenu as Mock;
const mockUseSisenseContext = useSisenseContext as Mock;

describe('DrilldownWidget', () => {
  const mockOpenMenu = vi.fn();
  const mockOnChange = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();

    // Provide default Sisense context
    mockUseSisenseContext.mockReturnValue({
      isInitialized: true,
      app: {
        defaultDataSource: DM.DataSource,
      },
      errorBoundary: {
        showErrorBox: true,
      },
    });

    mockUseMenu.mockReturnValue({ openMenu: mockOpenMenu });
  });

  it('renders child function and passes drilldown props (no menu opened by default)', () => {
    const { getByText } = render(
      <DrilldownWidget
        drilldownPaths={[DM.Category.Category, DM.Commerce.Gender]}
        onChange={mockOnChange}
        initialDimension={DM.Commerce.AgeRange}
      >
        {({ onDataPointsSelected }) => (
          <div>
            Child content
            <button onClick={() => onDataPointsSelected([], { clientX: 100, clientY: 200 } as any)}>
              Trigger
            </button>
          </div>
        )}
      </DrilldownWidget>,
    );

    expect(getByText('Child content')).toBeInTheDocument();

    // Clicking "Trigger" sets selected points but doesn't open the menu
    fireEvent.click(getByText('Trigger'));
    expect(mockOpenMenu).not.toHaveBeenCalled();
  });

  it('opens context menu if child calls onDataPointsSelected AND then calls onContextMenu with a position', () => {
    const { getByText } = render(
      <DrilldownWidget
        drilldownPaths={[DM.Category.Category]}
        initialDimension={DM.Commerce.Condition}
        onChange={mockOnChange}
      >
        {({ onDataPointsSelected, onContextMenu }) => {
          return (
            <button
              onClick={() => {
                onDataPointsSelected([], { clientX: 50, clientY: 60 } as any);
                onContextMenu({ top: 60, left: 50 });
              }}
            >
              Open Drilldown Menu
            </button>
          );
        }}
      </DrilldownWidget>,
    );

    fireEvent.click(getByText('Open Drilldown Menu'));

    // Now we expect openMenu to have been called
    expect(mockOpenMenu).toHaveBeenCalledTimes(1);
    expect(mockOpenMenu).toHaveBeenCalledWith(
      expect.objectContaining({
        position: { top: 60, left: 50 },
      }),
    );
  });

  it('renders breadcrumbs by default when there is a current drilldown dimension', () => {
    const { getByText, queryAllByText } = render(
      <DrilldownWidget
        drilldownPaths={[DM.Category.Category]}
        initialDimension={DM.Category.Category}
        onChange={mockOnChange}
        drilldownSelections={[{ points: [], nextDimension: DM.Category.Category }]}
      >
        {() => <div>With drilldown</div>}
      </DrilldownWidget>,
    );

    expect(getByText('With drilldown')).toBeInTheDocument();
    // find breadcrumbs
    expect(queryAllByText('Category (drilldown.breadcrumbsAllSuffix)')).toMatchSnapshot();
  });
});
