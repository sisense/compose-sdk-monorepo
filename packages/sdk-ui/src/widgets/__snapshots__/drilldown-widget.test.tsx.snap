// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DrilldownWidget > renders breadcrumbs by default when there is a current drilldown dimension 1`] = `
[
  <span
    class="mr-2"
  >
    Category (drilldown.breadcrumbsAllSuffix)
  </span>,
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1qfib76-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Category
     (
    drilldown.breadcrumbsAllSuffix
    )
    <span
      class="flex items-center justify-between"
    >
      <span
        class="csdk-absolute csdk-right-[0px] csdk-top-[3px] csdk-h-[18px] csdk-w-[18px] csdk-transform csdk-translate-x-1/2 csdk-rotate-45"
        style="background-color: transparent; border-top: 1px solid #F2F2F2; border-right: 1px solid #F2F2F2;"
      />
    </span>
  </button>,
]
`;
